### PR TITLE
Add random unitary matrices

### PR DESCRIPTION
**Pull request by philipdeboer**

### CODE REVIEW

**Comment by rgommers** [comment]
Thanks @philipdeboer!

@lesshaste would be much appreciated if you reviewed this

File: scipy/stats/tests/test_multivariate.py [hunk 1]
```diff
@@ -1403,6 +1404,57 @@ def test_to_corr(self):
...
+        xs = unitary_group.rvs(dim, size=samples)
+
+        eigs = np.vstack(scipy.linalg.eigvals(x) for x in xs)
+        x = np.arctan(eigs.imag/eigs.real)
+
+        h = np.histogram(x.flatten())[0]
+
+        # The angles "x" of the eigenvalues should be uniformly distributed
+        # We are not checking that the range is correct; maybe we should?
+        # Overall this seems to be a necessary but weak test of the distribution.
```
Comment (pv) [inline scipy/stats/tests/test_multivariate.py]: You can do a real statistical test, `res = kstest(samples, uniform(0, 2*pi).cdf); assert_(res.pvalue > 0.05)`.

File: scipy/stats/tests/test_multivariate.py [hunk 2]
```diff
@@ -1403,6 +1404,57 @@ def test_to_corr(self):
...
+        # the complex plane, are uncorrelated.
+
+        # Generate samples
+        dim = 2
+        samples = 10000  # Not too many, or the test takes too long
+        np.random.seed(514)  # Note that the test is sensitive to seed too
+        xs = unitary_group.rvs(dim, size=samples)
+
+        eigs = np.vstack(scipy.linalg.eigvals(x) for x in xs)
+        x = np.arctan(eigs.imag/eigs.real)
```
Comment (pv) [inline scipy/stats/tests/test_multivariate.py]: `np.arctan2(eigs.imag, eigs.real)`

File: scipy/stats/_multivariate.py [hunk 3]
```diff
@@ -3521,3 +3522,102 @@ def rvs(self, eigs, random_state=None, tol=1e-13, diag_tol=1e-7):
...
+    This class is similar to `ortho_group`.
+
+    References
+    ----------
+    .. [1] F. Mezzadri, "How to generate random matrices from the classical
+           compact groups", arXiv:math-ph/0609050v2.
+
+    Examples
+    --------
+    TODO Update this
```
Comment (pv) [inline scipy/stats/_multivariate.py]: TODO

File: scipy/stats/_multivariate.py [hunk 4]
```diff
@@ -3521,3 +3522,102 @@ def rvs(self, eigs, random_state=None, tol=1e-13, diag_tol=1e-7):
...
+        dim = self._process_parameters(dim)
+
+        random_state = self._get_random_state(random_state)
+
+        z = 1/math.sqrt(2)*(random_state.normal(size=(dim,dim)) +
+                            1j*random_state.normal(size=(dim,dim)))
+        q, r = scipy.linalg.qr(z)
+        d = r.diagonal()
+        d = d/abs(d)
+        q = q*d
```
Comment (pv) [inline scipy/stats/_multivariate.py]: `q *= d`

File: scipy/stats/_multivariate.py [hunk 5]
```diff
@@ -3521,3 +3522,102 @@ def rvs(self, eigs, random_state=None, tol=1e-13, diag_tol=1e-7):
...
+
+        dim = self._process_parameters(dim)
+
+        random_state = self._get_random_state(random_state)
+
+        z = 1/math.sqrt(2)*(random_state.normal(size=(dim,dim)) +
+                            1j*random_state.normal(size=(dim,dim)))
+        q, r = scipy.linalg.qr(z)
+        d = r.diagonal()
+        d = d/abs(d)
```
Comment (pv) [inline scipy/stats/_multivariate.py]: `d /= abs(d)`

File: scipy/stats/tests/test_multivariate.py [hunk 6]
```diff
@@ -1403,6 +1404,57 @@ def test_to_corr(self):
...
+        assert_raises(ValueError, unitary_group.rvs, 2.5)
+
+    def test_unitarity(self):
+        xs = [unitary_group.rvs(dim)
+              for dim in range(2,12)
+              for i in range(3)]
+
+        # Test that these are unitary matrices
+        for x in xs:
+            assert_array_almost_equal(np.dot(x, x.conj().T),
```
Comment (pv) [inline scipy/stats/tests/test_multivariate.py]: Prefer using  assert_allclose with explicitly specified atol.

**Review by pv — CHANGES_REQUESTED** [review]

File: scipy/stats/tests/test_multivariate.py (hunk 1 above)
Comment (lesshaste) [inline scipy/stats/tests/test_multivariate.py]: Won't that fail one in twenty times?

Comment (pv) [inline scipy/stats/tests/test_multivariate.py]: Fixed random seed. Of course, the p-threshold can be made smaller.

**Comment by philipdeboer** [comment]
Thanks for your feedback; I hope to get to this later this week.
Sent from my BlackBerry device on the Rogers Wireless Network

-----Original Message-----

**Comment by philipdeboer** [comment]
The uniform cdf confused me for a bit. I expected to be able to use uniform(-pi, pi).cdf, but that did not work. Is that a bug in uniform that I should fix? I had to "manually" shift the range to match my data. I guess I could have shifted my data...ah well.

**Comment by rkern** [comment]
You want a uniform distribution over the range `[-pi, +pi]`? The right syntax is `uniform(-pi, 2*pi)`. Like all of the distributions, `uniform()` takes a `loc` and `scale` parameter, so that's `uniform(loc=-pi, scale=2*pi)`. The two arguments are not the endpoints of the support.

File: scipy/stats/tests/test_multivariate.py [hunk 7]
```diff
@@ -1403,6 +1404,51 @@ def test_to_corr(self):
...
+        samples = 1000  # Not too many, or the test takes too long
+        np.random.seed(514)  # Note that the test is sensitive to seed too
+        xs = unitary_group.rvs(dim, size=samples)
+
+        # The angles "x" of the eigenvalues should be uniformly distributed
+        # Overall this seems to be a necessary but weak test of the distribution.
+        eigs = np.vstack(scipy.linalg.eigvals(x) for x in xs)
+        x = np.arctan2(eigs.imag, eigs.real)
+        res = kstest(x.ravel(), uniform(-np.pi, 2*np.pi).cdf)
+        assert_(res.pvalue > 0.8)
```
Comment (pv) [inline scipy/stats/tests/test_multivariate.py]: This maybe should be smaller threshold, `0.05` is one false rejection in 1/20, `0.01` 1/100, etc.
Although the larger value works for the particular seed above, the value probably should be smaller.

**Pull request merged**